[WIP] Eliminating guards with value range analysis.

ysiraichi opened the pull request:
This PR introduces a few simplification rules (suggested by @nunoplopes) to the guards expressions, and tries to eliminate redundant guards using the results of value range analysis. In summary the steps are:

1. Simplify the guard expressions
3. Filter those of the form `<symbol> <relop> <expr>`
4. Symbolically interpret `<expr>`, retrieving its value range
5. Update `<symbol>` value ranges according to `<relop>`
6. Eliminate the guards that cannot be decided by the updated value ranges
    - i.e. symbolically interpreting them results in a singleton value (e.g. true).

<details>
<summary>Issued Guards: resnet18</summary>

```python
# ============================================================
# = Before ===================================================
# ============================================================
x.size()[1] == 3
x.size()[3] == x.size()[2]
x.stride()[0] == 3*x.size()[2]**2
x.stride()[1] == x.size()[2]**2
x.stride()[2] == x.size()[2]
x.stride()[3] == 1
x.storage_offset() == 0
3*x.size()[2]**2 <= 2147483647
64*((x.size()[2] - 1)//2)**2 + 128*((x.size()[2] - 1)//2) + 64 <= 2147483647
Ne((x.size()[2] - 1)//2 + 1, 1)
Ne(((x.size()[2] - 1)//2)**2 + 2*((x.size()[2] - 1)//2) + 1, 1)
Ne((x.size()[2] - 1)//2 + 1, 1)
((x.size()[2] - 1)//2)**2 + 2*((x.size()[2] - 1)//2) + 1 > 1
Ne(Mod((x.size()[2] - 1)//2, 2), 0)
Ne((x.size()[2] - 1)//4 + 1, 1)
Ne(((x.size()[2] - 1)//4)**2 + 2*((x.size()[2] - 1)//4) + 1, 1)
64*((x.size()[2] - 1)//4)**2 + 128*((x.size()[2] - 1)//4) + 64 <= 2147483647
x.size()[0] < 16
((x.size()[2] - 1)//4)**2 + 2*((x.size()[2] - 1)//4) + 1 > 1
Ne((x.size()[2] - 1)//4 + 1, 1)
128*((x.size()[2] - 1)//8)**2 + 256*((x.size()[2] - 1)//8) + 128 <= 2147483647
Ne((x.size()[2] - 1)//8 + 1, 1)
Ne(((x.size()[2] - 1)//8)**2 + 2*((x.size()[2] - 1)//8) + 1, 1)
Ne((x.size()[2] - 1)//8 + 1, 1)
((x.size()[2] - 1)//8)**2 + 2*((x.size()[2] - 1)//8) + 1 > 1
256*((x.size()[2] - 1)//16)**2 + 512*((x.size()[2] - 1)//16) + 256 <= 2147483647
Ne((x.size()[2] - 1)//16 + 1, 1)
Ne(((x.size()[2] - 1)//16)**2 + 2*((x.size()[2] - 1)//16) + 1, 1)
Ne((x.size()[2] - 1)//16 + 1, 1)
((x.size()[2] - 1)//16)**2 + 2*((x.size()[2] - 1)//16) + 1 > 1
512*((x.size()[2] - 1)//32)**2 + 1024*((x.size()[2] - 1)//32) + 512 <= 2147483647
Ne((x.size()[2] - 1)//32 + 1, 1)
Ne(((x.size()[2] - 1)//32)**2 + 2*((x.size()[2] - 1)//32) + 1, 1)
Ne((x.size()[2] - 1)//32 + 1, 1)
((x.size()[2] - 1)//32)**2 + 2*((x.size()[2] - 1)//32) + 1 > 1
2 <= x.size()[0]
2 <= x.size()[2]


# ============================================================
# = After ====================================================
# ============================================================
x.size()[1] == 3
x.size()[3] == x.size()[2]
x.stride()[0] == 3*x.size()[2]**2
x.stride()[1] == x.size()[2]**2
x.stride()[2] == x.size()[2]
x.stride()[3] == 1
x.storage_offset() == 0
3*x.size()[2]**2 <= 2147483647
64*((x.size()[2] - 1)//2)**2 + 128*((x.size()[2] - 1)//2) <= 2147483583
Ne(Mod((x.size()[2] - 1)//2, 2), 0)
64*((x.size()[2] - 1)//4)**2 + 128*((x.size()[2] - 1)//4) <= 2147483583
128*((x.size()[2] - 1)//8)**2 + 256*((x.size()[2] - 1)//8) <= 2147483519
256*((x.size()[2] - 1)//16)**2 + 512*((x.size()[2] - 1)//16) <= 2147483391
512*((x.size()[2] - 1)//32)**2 + 1024*((x.size()[2] - 1)//32) <= 2147483135
x.size()[2] >= 33
2 <= x.size()[0]
```

</details>

@ezyang

ysiraichi committed:
First version of guards simplification.

ysiraichi commented:
A few more numbers:

| Benchmark | Guards (before) | Guards (After) |
|---------------|--------------|---------------------|
| Background_Matting | 36/**69** | 36/**61** |
| LearningToPaint | 9/**52** | 9/**23** |
| attention_is_all_you_need_pytorch | 10/**32** | 10/**28** |
| dcgan | 9/**71** | 9/**26** |
| dlrm | 34/**48** | 34/**34** |
| drq | 9/**84**/**144** | 9/**50**/**110** |
| maml_omniglot | 9/**106** | 9/**45** |
| resnet152 | 9/**44** | 9/**23** |
| resnet18 | 9/**37** | 9/**16** |
| resnet50 | 9/**44** | 9/**23** |
| functorch_maml_omniglot | 9/**109** | 9/**47** |
| pytorch_CycleGAN_and_pix2pix | 9/**33** | 9/**23** |
| pytorch_stargan | 14/**81** | 14/**37** |
| pytorch_unet | 9/**635**/**635** | 9/**268**/**268** |
| vgg16 | 9/**83** | 9/**26** |

Notes: 
- The '/' separate guards for each generated graph
- I still haven't run them all (will do so next)

ezyang commented:
The guard reduction stats look like bangers. Sweet!

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1721,6 +1806,15 @@ def _update_divisible(self):
 
     @_lru_cache
     def simplify(self, expr: "sympy.Expr") -> "sympy.Expr":
+        def get_const(expr):
```
This name is super undescriptive and there's no doc either....

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1755,6 +1849,34 @@ def simplify(self, expr: "sympy.Expr") -> "sympy.Expr":
+                    and lhs_const != 0
+            ):
+                expr = type(expr)(expr.lhs - lhs_const, expr.rhs - lhs_const)  # type: ignore
+        if isinstance(expr, sympy.Eq) and isinstance(expr.lhs, FloorDiv):
```
Should we always do this simplification even if the rhs is a symbol?

ysiraichi replied:
It helps us get rid of `FloorDiv`. But, probably we shouldn't do this if there are some simplification rules that depend on that operation. I will move this out.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1550,10 +1550,64 @@ def _verify(expr, potential_expr):
-            if self._maybe_evaluate_static(g) is not None:
+        def simplify_until_no_change(expr: sympy.Expr) -> sympy.Expr:
+            new_expr = self.simplify(expr)
+            while new_expr != expr:
```
we should also have a max number of iterations that we are willing to do here. Also, we need to audit the preexisting rules to make sure they're terminating, and we need to update the invariants for simplify to say that they need to be terminating.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1550,10 +1550,64 @@ def _verify(expr, potential_expr):
+            assert symbol in self.var_to_range
+
+            # Consider actually issuing the guard if it is used for refining
+            # the value range of a symbol. Otherwise, it is discarded.
```
Refining the ranges at the end seems worse. You want to update the ranges AS SOON as you know you have a refined range! Do it earlier, it will save work!

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1550,10 +1550,64 @@ def _verify(expr, potential_expr):
+            if self._maybe_evaluate_static_impl(g, var_to_range=var_to_range) is not None:
+                continue
+
+            if not isinstance(g, sympy.Rel):
+                unhandled_guards.append((g, tb))
+                continue
+
+            symbol = g.lhs
+
+            if not isinstance(g.lhs, sympy.Symbol):
+                unhandled_guards.append((g, tb))
                 continue
```
We don't want to call solve, it is too likely to infinite loop.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1550,10 +1550,64 @@ def _verify(expr, potential_expr):
+                upper = rhs_vr.upper - (1 * isinstance(g, sympy.Lt))
+                upper_guard = (g, tb)
+
+            # If the value range is still the same, ignore the guard.
```
This seems wrong.

I can have a guard which doesn't improve my range at all, but is still required; e.g., a guard that x is even doesn't refine range, but you can't drop it.

ysiraichi replied:
You are right. I guess the safe choice here is to let `_maybe_evaluate_static` decide.

ysiraichi committed:
Move value range refinement to `evaluate_expr`.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1550,10 +1550,64 @@ def _verify(expr, potential_expr):
+
+        simplified_guards = [(simplify_until_no_change(g), tb) for g, tb in self.guards]
+        var_to_guards = dict()
+        var_to_range = self.var_to_range.copy()
```
When you refine ranges, you should modify var_to_range directly. However, I would have expected we'd want to refine ranges *as we are adding guards*, not all the way at the end.

ysiraichi commented:
@ezyang
I have moved the value range refinement to the end of `evaluate_expr` function (just after adding the guard). In `produce_guards` we:

1. try to eliminate the guards by calling `_maybe_evaluate_static`
2. add the guards that refine the value ranges stored in: `self.var_to_guards`

I have re-run a few TorchBench benchmarks (those from my previous comment). See the results below. 
Next, I will run the whole TorchBench for completeness.

-----

Apparently, I was adding many duplicated guards that were coming from `self.guards` (e.g. `pytorch_unet` had 458 duplicated guards).

For comparison, I added a column that showed the number of guards _if we just eliminated the duplicated guards_.  In summary, this value range guard elimination complements the duplicated guards removal.

| Benchmark                         | Guards (before)   | Guards (dedup)   | Guards (After)   |
|-----------------------------------|-------------------|------------------|------------------|
| Background_Matting                | 36/**69**         | 36/**67**        | 36/**61**        |
| LearningToPaint                   | 9/**52**          | 9/**43**         | 9/**21**         |
| attention_is_all_you_need_pytorch | 10/32             | 10/32            | 10/32            |
| dcgan                             | 9/**71**          | 9/**64**         | 9/**27**         |
| dlrm                              | 34/**48**         | 34/**42**        | 34/**36**        |
| drq                               | 9/**84**/**144**  | 9/**71**/**131** | 9/**49**/**109** |
| maml_omniglot                     | 9/**106**         | 9/**87**         | 9/**43**         |
| resnet152                         | 9/**44**          | 9/**39**         | 9/**24**         |
| resnet18                          | 9/**37**          | 9/**32**         | 9/**17**         |
| resnet50                          | 9/**44**          | 9/**39**         | 9/**24**         |
| functorch_maml_omniglot           | 9/**109**         | 9/**90**         | 9/**45**         |
| pytorch_CycleGAN_and_pix2pix      | 9/**33**          | 9/**29**         | 9/**23**         |
| pytorch_stargan                   | 14/**81**         | 14/**77**        | 14/**36**        |
| pytorch_unet                      | 9/**635**/**635** | 9/**177**        | 9/**73**/**73**  |
| vgg16                             | 9/**83**          | 9/**69**         | 9/**27**         |

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1212,6 +1213,7 @@ def __init__(
         self.var_to_range: Dict["sympy.Symbol", ValueRanges] = {}
         self.var_to_sources: Dict["sympy.Symbol", List[Source]] = {}
         self.var_to_stack: Dict["sympy.Symbol", str] = {}
+        self.var_to_guards: Dict["sympy.Symbol", Tuple[Optional[ShapeGuard], Optional[ShapeGuard]]] = {}
```
A little comment explaining what this is would be good.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1920,14 +1987,76 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
+        it = 0
+        expr = guard.expr
+
+        while True:
```
nit: `while _ in range(max_iterations)`

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1920,14 +1987,76 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
+
+        # Filter the guards that:
+        #   1. are relational operations
+        #   2. have a symbol as the left-hand side
```
Can the symbol show up on rhs or is that impossible?

ysiraichi replied:
I believe they can. So, here, we are being restrictive: `maybe_simplify_floordiv` tries to move stuff to the right so that we are left with a symbol in the left.

ysiraichi replied:
We could, however, invert the expression (e.g. `a < b` into `b > a`) and re-run the value refinement procedure again. How does this sound?

ysiraichi replied:
I have implemented this idea. Let me know what you think.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1719,8 +1741,43 @@ def _update_divisible(self):
         self.divisible = new_divisible
 
+    @_lru_cache
+    def maybe_simplify_floordiv(self, expr: "sympy.Expr") -> "sympy.Expr":
```
TODO: review this for soundness

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1920,14 +1987,76 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
+        lower_guard, upper_guard = self.var_to_guards.get(symbol, (None, None))
+
+        if lower < rhs_vr.lower and isinstance(expr, (sympy.Eq, sympy.Ge, sympy.Gt)):
+            lower = rhs_vr.lower + (1 * isinstance(expr, sympy.Gt))
```
nit: maybe clearer to do int(isinstance(...)) instead?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1920,14 +1987,76 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
+
+
+        # Update the value range of the left-hand side, if the
+        # right-hand side provides a better range.
```
oh looks like you're doing some sort of inner VR analysis, sounds good

ysiraichi replied:
A summary of what is going on:
1. Filter expressions of form: `<symbol> <relop> <expr>`
2. Evaluate the value range of: `<expr>`
3. Update `<symbol>` value range if the one from (2) is better (lower or upper bound)
4. Clear the cache for `_maybe_evaluate_static`

> As @avikchaudhuri was pointing out in chat, we should only do value range updates for univariate formulas

Why? If there are any free variables on the RHS, `sympy_interp` would just get their value range, wouldn't it? I can't see why that would give incorrect results.

> I think this would eliminate the need for `var_to_guards` as well.

I don't think so. As far as I understand, we must issue the guards that were used for refining the value ranges, right? In that sense, a value range's lower and upper bound might be refined by different guards. Thus, we need to remember what are those guards.

ezyang replied:
OK, so I agree that this is not obviously unsound (what I suggested was a more conservative version of this, your PR is going even further.) However, I am still struggling to find a simple explanation for the refinement procedure we're doing here. I'll give more specific comments inline.

lezcano replied:
Note that point 2 and 3 here are just: if `x < f(t)` for t in som erange, and for those t we have that `a < f(t) < b` then since `<` is transitive` x < b` irrespective of `t`. Note that this simple argument also goes through if instead of `f(t)` we have `f(t1, t2, ..., tn)`. This may help with `where`, `mul` or `div` clauses.

ysiraichi committed:
Mirror expression and re-run refinement.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1920,14 +1990,92 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
+            rhs_vr = sympy_interp(ValueRangeAnalysis, self.var_to_range, expr.rhs)  # type: ignore
+            lower_guard, upper_guard = self.var_to_guards.get(symbol, (None, None))
+
+            if lower < rhs_vr.lower and isinstance(expr, (sympy.Eq, sympy.Ge, sympy.Gt)):
```
I suppose a comment here would make the bounds here clearer. Something like "Let's suppose that we have a preexisting range for x `[0, 100]`. Now, we issue a guard `x > y`, where the range for y is `[50, 150]`. Then, `lower = 0, rhs_vr.lower = 50` and therefore refinement can happen, refining x to `[50, 100]`, since x must be greater than y, but the lowest y could be is 50."

ezyang commented:
We're going to need to come up with some sort of testing strategy, since if we get the simplification rules wrong nothing will tell us, we will be silently and subtly wrong.  One easy-ish thing to do is in a debug mode, accumulate guards WITHOUT any simplification, and when we pass a guard check, run through all of them to ensure that they are all satisfied. (I suppose we could also run through all of them when we fail a guard check too.) This will require some nontrivial Dynamo extra scaffolding.

But for the logic here, hmm, a good start would be to write manual test cases that trigger all of the branches in the logic here. That will at least give us confidence that it is not horribly wrong. Another possibility is to beef up debug checking inside the simplification code itself. For example, after we refine ranges, we can test some randomly sample some points (maybe deterministically sample points on the boundary of the ranges) and check that the ranges are consistent. IDK if @nunoplopes has some other ideas.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1920,14 +1990,92 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
+                return
+
+            self.var_to_range[symbol] = ValueRanges(lower, upper)
+            self.var_to_guards[symbol] = (lower_guard, upper_guard)
```
We should setup some logging here. @avikchaudhuri do you have a preference?

ysiraichi commented:
> One easy-ish thing to do is in a debug mode, accumulate guards WITHOUT any simplification

I think I can do this by:
- make `_maybe_evaluate_static` use a custom range dictionary
    - at first, inside `evaluate_expr`, call it with the original ranges
    - then, inside `produce_guards`, call it with the refined ranges if the debug parameter (new) is set
- [call `produce_guards`](https://github.com/pytorch/pytorch/blob/4c0dce50fd002d59c9f2ad85dcf57e7fbfb05071/torch/_dynamo/guards.py#L417-L431) twice if on debug mode, alternating the value of the debug parameter
- generate the code for comparing the simplified and original guards

@ezyang what do you think?

ysiraichi committed:
Add comment explaining refinement.

ezyang commented:
I don't think that's enough, since we will refine variables when we discover they are constants by putting them in replacements. Actually, I suppose those are not too bad, because it's easy to check if a refinement is compatible (since you had better refine to the hinted value), but maybe you can trigger something more obscure, like adding an extra divisibility constraint as the result of an unsound simplification.

For asserts, you really want it to be dog obvious that it is sound. In this case, it seems to me like the most obvious thing is to do literally no symbolic reasoning. Maybe easiest to setup two ShapeEnvs in parallel for this.

nunoplopes commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1719,8 +1744,43 @@ def _update_divisible(self):
 
+    @_lru_cache
+    def maybe_simplify_floordiv(self, expr: "sympy.Expr") -> "sympy.Expr":
+        if isinstance(expr, sympy.Eq) and isinstance(expr.lhs, FloorDiv):
```
this bit is correct for positive numbers, but not for negatives. Is that important?  I guess not?
For example, take rhs = 1, lhs = -1 / -1.
The 2nd expression becomes -1 - (2 * -1) < 0  <=>  1 < 0  <=>  false.

For the positive case, we have this smt file:
```smt
(declare-const rhs (_ BitVec 8))
(declare-const numerator (_ BitVec 8))
(declare-const denominator (_ BitVec 8))

; don't divide by 0
(assert (not (= denominator #x00)))

; non-negative
(assert (bvsge rhs #x00))
(assert (bvsge numerator #x00))
(assert (bvsge denominator #x00))
; disallow overflows by restricting vars to 4 bits, and intermediates to 8
(assert (bvsle rhs #x0f))
(assert (bvsle numerator #x0f))
(assert (bvsle denominator #x0f))

(assert (not (=
  (= (bvsdiv numerator denominator) rhs)
  (and (bvsge (bvsub numerator (bvmul rhs denominator)) #x00)
       (bvslt (bvsub numerator (bvmul (bvadd rhs #x01) denominator)) #x00)
  )
)))

(check-sat)
(get-model)
```

ezyang replied:
Size compute is usually positive, but in principle it is possible for intermediate expressions to go negative. So best to do this properly.

lezcano replied:
When do we have a substraction (or a negation) in a size computation?

peterbell10 replied:
A concrete example is `torch.arange(100, 0, -1, dtype=torch.int64)`. The size is `-100 // -1`

ysiraichi replied:
Not sure, but I found some guards like: `s3//8 - 2*s3//16 < 0`

ysiraichi replied:
I have fixed this by making sure the denominator is always a positive integer constant.

nunoplopes commented:
> We're going to need to come up with some sort of testing strategy, since if we get the simplification rules wrong nothing will tell us, we will be silently and subtly wrong. One easy-ish thing to do is in a debug mode, accumulate guards WITHOUT any simplification, and when we pass a guard check, run through all of them to ensure that they are all satisfied. (I suppose we could also run through all of them when we fail a guard check too.) This will require some nontrivial Dynamo extra scaffolding.
> 
> But for the logic here, hmm, a good start would be to write manual test cases that trigger all of the branches in the logic here. That will at least give us confidence that it is not horribly wrong. Another possibility is to beef up debug checking inside the simplification code itself. For example, after we refine ranges, we can test some randomly sample some points (maybe deterministically sample points on the boundary of the ranges) and check that the ranges are consistent. IDK if @nunoplopes has some other ideas.

Agreed.
A sort of translation validation would be great. You just need to check that the guards before and after simplification are equivalent.
The problem is that sympy can't do that equivalence check, you would need Z3. And then we are back to the problem of do you want to go all in on Z3, do you want to have a translation between sympy and Z3 expressions and run translation validation only when Z3 is enabled.

Generating tests is also good, but since this code requires a sequence of bounds, it requires a lot of tests to trigger all cases. It's unlikely our generator would cover all cases. Hence the usefulness of the translation validation.

ezyang commented:
OK, I'm sold on translation validation. At least to start, let's keep Z3 optional (so still doing the Sympy stuff), but having a working translation validator would make it easier to go all in on Z3 if we decide to.

ysiraichi committed:
Fix conditions.

ezyang commented:
If we can do a simple univariate range analysis, I might be willing to put it in before we have translation validation.

However, I am *already* running into cases where our existing logic, as is, is failing translation validation. (A recent example: when we have `s1*s2*s3 = s3*s4*s4`, where s3 is not known to be positive, we will incorrectly refine `s3 = 0`, even though another valid solution was to set `s4=s1=s2`.) So I think I would prefer to do translation validation first.

ezyang commented:
See also https://github.com/pytorch/pytorch/pull/97792 where a very simple translation validation is identifying broken simplification rules in ShapeEnv
